src/main/java/com/patchworkmc/event/EventHandlerRewriter.java

```diff
@@ -12,37 +12,54 @@
 import org.objectweb.asm.Type;
 
 import com.patchworkmc.Patchwork;
-
-public class EventHandlerScanner extends ClassVisitor {
-	private static final String REGISTER_STATIC = "patchwork_registerStaticEventHandlers";
-	private static final String REGISTER_INSTANCE = "patchwork_registerInstanceEventHandlers";
-	private static final String METAFACTORY_OWNER = "java/lang/invoke/LambdaMetafactory";
-	private static final String METAFACTORY_DESCRIPTOR = "(Ljava/lang/invoke/MethodHandles$Lookup;Ljava/lang/String;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodType;Ljava/lang/invoke/MethodHandle;Ljava/lang/invoke/MethodType;)Ljava/lang/invoke/CallSite;";
-	private static final String EVENT_BUS = "net/minecraftforge/eventbus/api/IEventBus";
-
-	private Consumer<EventBusSubscriber> subscriberConsumer;
-	private Consumer<SubscribeEvent> subscribeEventConsumer;
-	private List<SubscribeEvent> staticSubscribers;
+import com.patchworkmc.Constants.EventBus;
+import com.patchworkmc.Constants.Lambdas;
+
+public class EventHandlerRewriter extends ClassVisitor {
+	private final Consumer<EventBusSubscriber> subscriberConsumer;
+	private final Consumer<SubscribeEvent> subscribeEventConsumer;
+	// boolean is "isStatic"
+	private final HashMap<SubscribeEvent, Boolean> subscribeEvents = new HashMap<>();
 	private String className;
+	private boolean hasStatic = false;
+	private boolean hasInstance = false;
+
+	private boolean isInterface = false;
+	private boolean isFinalClass = false;
 
-	public EventHandlerScanner(ClassVisitor parent, Consumer<EventBusSubscriber> subscriberConsumer, Consumer<SubscribeEvent> subscribeEventConsumer) {
+	public EventHandlerRewriter(ClassVisitor parent, Consumer<EventBusSubscriber> subscriberConsumer, Consumer<SubscribeEvent> subscribeEventConsumer) {
```

Comment:
`subscribeEventConsumer` is only used to determine whether a class has static or instance event handlers in `PatchworkTransformer`, but that's already done here with `hasStatic` and `hasInstance`. Maybe those flags could be exposed somehow? Probably not in scope for this PR, but it would be a nice change for the future.

Reply:
I think this refactor would be better suited as part of cleaning up the awful PatchworkTransformer#accept method

Reply:
actually, screw it, i'll clean it up a bit

Reply:
I had to leave it in for mod ATs (which are a bit of a mess anyways).
I went ahead and avoided using it for anything other than the AT to make it easier to remove later.